url74b2be4e76_gurbano_petulant_octo_sudoku_tgz-pJ8-test_testJ8/src/ocr/Entry.java

```diff
@@ -0,0 +1,356 @@
+    * crop the image so that whitespace is
+    * removed.
+    *
+    * @param w The width of the image.
+    * @param h The height of the image
+    */
+   protected void findBounds(int w,int h)
+   {
+     // top line
+     for ( int y=0;y<h;y++ ) {
+       if ( !hLineClear(y) ) {
+         downSampleTop=y;
+         break;
+       }
+
+     }
+     // bottom line
+     for ( int y=h-1;y>=0;y-- ) {
+       if ( !hLineClear(y) ) {
+         downSampleBottom=y;
+         break;
+       }
+     }
+     // left line
+     for ( int x=0;x<w;x++ ) {
+       if ( !vLineClear(x) ) {
+         downSampleLeft = x;
+         break;
+       }
+     }
+
+     // right line
+     for ( int x=w-1;x>=0;x-- ) {
+       if ( !vLineClear(x) ) {
+         downSampleRight = x;
+         break;
+       }
+     }
+   }
+
+   /**
+    * Called to downsample a quadrant of the image.
+    *
+    * @param x The x coordinate of the resulting
+    * downsample.
+    * @param y The y coordinate of the resulting
+    * downsample.
+    * @return Returns true if there were ANY pixels
+    * in the specified quadrant.
+    */
+   protected boolean downSampleQuadrant(int x,int y)
+   {
+     int w = entryImage.getWidth(this);
+     int startX = (int)(downSampleLeft+(x*ratioX));
+     int startY = (int)(downSampleTop+(y*ratioY));
+     int endX = (int)(startX + ratioX);
+     int endY = (int)(startY + ratioY);
+
+     for ( int yy=startY;yy<=endY;yy++ ) {
+       for ( int xx=startX;xx<=endX;xx++ ) {
+         int loc = xx+(yy*w);
+
+         if ( pixelMap[ loc  ]!= -1 )
+           return true;
+       }
+     }
+
+     return false;
+   }
+
+
+   /**
+    * Called to downsample the image and store
+    * it in the down sample component.
+    */
+   public void downSample()
+   {
+     int w = entryImage.getWidth(this);
+     int h = entryImage.getHeight(this);
+
+     PixelGrabber grabber = new PixelGrabber(
+                                            entryImage,
+                                            0,
+                                            0,
+                                            w,
+                                            h,
+                                            true);
+     try {
+
+       grabber.grabPixels();
+       pixelMap = (int[])grabber.getPixels();
+       findBounds(w,h);
+
+       // now downsample
+       SampleData data = sample.getData();
+
+       ratioX = (double)(downSampleRight-
+                         downSampleLeft)/(double)data.getWidth();
+       ratioY = (double)(downSampleBottom-
+                         downSampleTop)/(double)data.getHeight();
+
+       for ( int y=0;y<data.getHeight();y++ ) {
+         for ( int x=0;x<data.getWidth();x++ ) {
+           if ( downSampleQuadrant(x,y) )
+             data.setData(x,y,true);
+           else
+             data.setData(x,y,false);
+         }
+       }
+
+       sample.repaint();
+       repaint();
+     } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)